### PR TITLE
A faster method to evaluate symbolic targets coming from jump tables.

### PR DESCRIPTION
**Pull request by ltfish**
With this commit, we can easily support 2048 symbolic jump targets on my laptop without overloading Z3.

### CODE REVIEW

**Comment by ltfish** [comment]
@rhelmot RFC.

File: angr/engines/successors.py [hunk 1]
```diff
@@ -368,10 +369,98 @@ def _finalize(self):
...
+            ip_ = ip_.args[0]
+            outer_reverse = True
+
+        fallback = False
+        atom = None
+        concretes = set()
+        while True:
+            # We must fully unpack the entire AST to make sure it indeed complies with the form above
+            if ip_.op == "If":
+                cond = ip_.args[0]
```
Comment (rhelmot) [inline angr/engines/successors.py]: what happens if the args order is reversed?

Comment (rhelmot) [reply angr/engines/successors.py]: wait, this is an if, not an eq

**Comment by rhelmot** [comment]
It looks fine to me but it's very brittle with respect to the exact structure of the tree. It would be nice to write something to convert an ITE tree to its cases more generally.

**Comment by rhelmot** [comment]
yours only works if the tree is linear, descending to the left, and has the equalities with their operations in a certain order

**Comment by rhelmot** [comment]
but you also can't possibly say that the recursive way isn't WAY more maintainable :P

**Comment by ltfish** [comment]
> yours only works if the tree is linear, descending to the left, and has the equalities with their operations in a certain order
 
The “descending to the left” part should be fixed. It is intentional to only support equalities, because otherwise there is no way for us to make sure the conditions are mutually exclusive.

**Comment by ltfish** [comment]
> but you also can't possibly say that the recursive way isn't WAY more maintainable :P
 
It definitely is! But I hit a stack overflow with one of my test cases :(

**Comment by rhelmot** [comment]
I'll clean it up to use less memory and add it to claripy along with the And simplifier.

**Comment by rhelmot** [comment]
Wait a second though, why? Nobody should have created an ITE tree that deep...

**Comment by ltfish** [comment]
> Wait a second though, why? Nobody should have created an ITE tree that deep...
 
How else are we going to support a jump table with 32k targets?

**Comment by ltfish** [comment]
This is ready to be merged in - the only failure comes from the API coverage detection in angr-docs. @rhelmot @zardus comments? thoughts?

**Comment by ltfish** [comment]
> Furthermore a whole plugin seems like a bad idea - I'd like that issue to be solved on the same level as `options`

I'm fine with either way. Should I reimplement the state options as a whole new class, which supports both boolean switches and key-value settings?

**Comment by rhelmot** [comment]
That was what I had in mind :) ideally it would subclass dict and implement
the interface of set too by just setting True whenever an option was added
like that.

**Comment by ltfish** [comment]
I will merge this in, and then implement a `SimStateOptions` class.

**Comment by rhelmot** [comment]
sounds good to me.

**Pull request merged**